packages/schematics/angular/migrations/update-9/update-server-main-file.ts

```diff
@@ -0,0 +1,136 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import { Rule } from '@angular-devkit/schematics';
+import * as ts from '../../third_party/github.com/Microsoft/TypeScript/lib/typescript';
+import { findNodes } from '../../utility/ast-utils';
+import { findPropertyInAstObject } from '../../utility/json-utils';
+import { Builders } from '../../utility/workspace-models';
+import { getTargets, getWorkspace } from './utils';
+
+/**
+ * Update the `main.server.ts` file by adding exports to `renderModule` and `renderModuleFactory` which are
+ * now required for Universal and App-Shell for Ivy and `bundleDependencies`.
+ */
+export function updateServerMainFile(): Rule {
+  return tree => {
+    const workspace = getWorkspace(tree);
+
```

Comment:
Please add comments to individual sections

- Find the server entry point
- Find if there are existing exports of renderModule or renderModuleFactory

etc.

Reply:
added